google/cloud/pubsub/internal/tracing_message_batch.cc

```diff
@@ -40,8 +40,40 @@ void TracingMessageBatch::SaveMessage(pubsub::Message m) {
   child_->SaveMessage(std::move(m));
 }
 
-// TODO(#12528): Implement functionality for Flush.
-void TracingMessageBatch::Flush() { child_->Flush(); }
+void TracingMessageBatch::Flush() {
+  using opentelemetry::trace::SpanContext;
+  using AttributesList =
+      std::vector<std::pair<opentelemetry::nostd::string_view,
+                            opentelemetry::common::AttributeValue>>;
+  int64_t message_count = message_spans_.size();
+  auto constexpr kMaxOtelLinks = 128;
+  std::vector<std::pair<SpanContext, AttributesList>> links;
+
+  // If the batch size is less than the max size, add the links to a single
+  // span.
+  if (message_count < kMaxOtelLinks) {
+    for (int64_t i = 0; i < message_count; i++) {
+      auto span = message_spans_.at(static_cast<int>(i));
+      AttributesList link_attributes = {{"messaging.pubsub.message.link", i}};
+      links.emplace_back(std::make_pair(span->GetContext(), link_attributes));
+    }
```

Comment:
Consider:

```suggestion
    std::int64_t i = 0;
    for (auto const& span : message_spans_) {
      AttributesList link_attributes = {{"messaging.pubsub.message.link", i++}};
      links.emplace_back(std::make_pair(span->GetContext(), link_attributes));
    }
```

Reply:
And then we realize that this is a transform, and we should do:

```cc
std::transform(message_spans_.begin(), message_spans_.end(), std::back_inserter(links),
    [i = std::int64_t(0)](auto const& span) mutable {
       return std::make_pair(span->getContext(), AttributeList{{"messaging.pubsub.message.link", i++}});
    });
```

That will require a couple of extra `#include` directives.